<a href="https://colab.research.google.com/github/davseg/hacking-civico/blob/master/tareas/03_exploracion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p align="center">
<img src="http://codeandomexico.org/resources/img/codeandomexico.png" width="500" alt="Codeando México"><br>
<a href="http://www.codeandomexico.org/" target="_blank"><img src="https://img.shields.io/badge/website-CodeandoMexico-00D88E.svg"></a>
<a href="http://slack.codeandomexico.org/" target="_blank"><img src="https://img.shields.io/badge/slack-CodeandoMexico-EC0E4F.svg"></a>
</p>
<!-- __ -->

# Curso de Datos Abiertos y Hacking Cívico

Este curso tiene como objetivo habilitar las capacidades de la ciudadanía y los servidores públicos en el uso y generación de datos abiertos para el bien común. Puedes encontrar más información [aquí](https://github.com/CodeandoMexico/hacking-civico).

# Exploración de datos: COVID-19 en México

En este cuaderno aprenderemos algunas bases de exploración de datos. Los datos a utilizar son los datos abiertos sobre COVID-19 en México, los cuales pueden ser accedidos directamente desde el portal: https://www.gob.mx/salud/documentos/datos-abiertos-152127

Del portal de datos abiertos podemos encontrar dos enlaces, uno para descargar el conjunto de datos y otro para descargar el diccionario; así que podemos automatizar el proceso de descarga y descompresión de la información.

Para obtener los datos, sólo basta que ejecutes la siguiente línea de código, que contiene el comando de descarga de la información.

In [30]:
# Para descargar el conjunto de datos
!wget http://datosabiertos.salud.gob.mx/gobmx/salud/datos_abiertos/datos_abiertos_covid19.zip

# Para descargar el diccionario de los datos
!wget http://epidemiologia.salud.gob.mx/gobmx/salud/datos_abiertos/diccionario_datos_covid19.zip

# Descomprimimos ambas carpetas y movemos archivos
!unzip datos_abiertos_covid19.zip
!unzip diccionario_datos_covid19.zip
!mv ./diccionario_datos_covid19/*.xlsx .
!mv *COVID19MEXICO.csv COVID19MEXICO.csv

# Eliminamos las carpetas comprimidas
!rm -rf diccionario_datos_covid19
!rm datos_abiertos_covid19.zip
!rm diccionario_datos_covid19.zip

--2020-09-13 03:38:53--  http://datosabiertos.salud.gob.mx/gobmx/salud/datos_abiertos/datos_abiertos_covid19.zip
Resolving datosabiertos.salud.gob.mx (datosabiertos.salud.gob.mx)... 187.210.186.146
Connecting to datosabiertos.salud.gob.mx (datosabiertos.salud.gob.mx)|187.210.186.146|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 31945189 (30M) [application/x-zip-compressed]
Saving to: ‘datos_abiertos_covid19.zip’

datos_abiertos_covi 100%[===================>]  30.46M  17.0MB/s    in 1.8s    

2020-09-13 03:38:55 (17.0 MB/s) - ‘datos_abiertos_covid19.zip’ saved [31945189/31945189]

--2020-09-13 03:38:55--  http://epidemiologia.salud.gob.mx/gobmx/salud/datos_abiertos/diccionario_datos_covid19.zip
Resolving epidemiologia.salud.gob.mx (epidemiologia.salud.gob.mx)... 187.191.75.207
Connecting to epidemiologia.salud.gob.mx (epidemiologia.salud.gob.mx)|187.191.75.207|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ep

Hasta este punto, si abres la barra lateral izaquiera, deberías ser capaz de encontrar los datos:

- `COVID19MEXICO.csv`
- `Catalogos_0412.xlsx`
- `Descriptores_0419.xlsx`

👈🏼

## Carga

Ahora que tenemos los datos, procederemos a cargarlos. Para cargar los datos, primero importaremos algunos paquetes de Python, pues su funcionalidad nos será muy util.

In [31]:
import statistics

import pandas as pd
import numpy as np

In [32]:
# Especificamos el nombre del archivo por abrir
nombre_archivo = 'COVID19MEXICO.csv'

# Utilizamos la función de Pandas para cargar un archivo CSV
data = pd.read_csv(nombre_archivo, encoding='latin-1')

Mostramos la cabeza (primeros elementos del archivo).

In [33]:
data.head()

FECHA_ACTUALIZACION ID_REGISTRO  ORIGEN  ...  PAIS_NACIONALIDAD  PAIS_ORIGEN  UCI
0          2020-09-12      114cd3       2  ...            México           99   97
1          2020-09-12      0bafd8       2  ...            México           99    2
2          2020-09-12      18747a       2  ...            México           99   97
3          2020-09-12      174174       2  ...            México           99   97
4          2020-09-12      0eae86       2  ...            México           99    1

[5 rows x 35 columns]

Ahora procedemos a cargar el archivo con los descriptores.

In [34]:
nombre_diccionario = 'Descriptores_0419.xlsx'
diccionario = pd.read_excel(nombre_diccionario)

In [35]:
diccionario

Nº  ...                                   FORMATO O FUENTE
0    1  ...                                         AAAA-MM-DD
1    2  ...                                             TEXTO 
2    3  ...  CATÁLOGO: ORIGEN                              ...
3    4  ...  CATÁLOGO: SECTOR                              ...
4    5  ...                                CATALÓGO: ENTIDADES
5    6  ...       CATÁLOGO: SEXO                              
6    7  ...                                CATALÓGO: ENTIDADES
7    8  ...                                CATALÓGO: ENTIDADES
8    9  ...                               CATALÓGO: MUNICIPIOS
9   10  ...                           CATÁLOGO: TIPO_PACIENTE 
10  11  ...                                         AAAA-MM-DD
11  12  ...                                         AAAA-MM-DD
12  13  ...                                         AAAA-MM-DD
13  14  ...  CATÁLOGO: SI_ NO                              ...
14  15  ...  CATÁLOGO: SI_ NO                              ...
15  16  ...                                   NÚMERICA EN AÑOS
16  17  ...                            CATÁLOGO: NACIONALIDAD 
17  18  ...  CATÁLOGO: SI_ NO                              ...
18  19  ...  CATÁLOGO: SI_ NO                              ...
19  20  ...  CATÁLOGO: SI_ NO                              ...
20  21  ...  CATÁLOGO: SI_ NO                              ...
21  22  ...  CATÁLOGO: SI_ NO                              ...
22  23  ...  CATÁLOGO: SI_ NO                              ...
23  24  ...  CATÁLOGO: SI_ NO                              ...
24  25  ...  CATÁLOGO: SI_ NO                              ...
25  26  ...  CATÁLOGO: SI_ NO                              ...
26  27  ...  CATÁLOGO: SI_ NO                              ...
27  28  ...  CATÁLOGO: SI_ NO                              ...
28  29  ...  CATÁLOGO: SI_ NO                              ...
29  30  ...  CATÁLOGO: SI_ NO                              ...
30  31  ...                                CATÁLOGO: RESULTADO
31  32  ...  CATÁLOGO: SI_ NO                              ...
32  33  ...                               TEXTO, 99= SE IGNORA
33  34  ...                               TEXTO, 97= NO APLICA
34  35  ...  CATÁLOGO: SI_ NO                              ...

[35 rows x 4 columns]

Finalmente, cargamos el rachivo de catálogos.

In [36]:
nombre_cat = 'Catalogos_0412.xlsx'
sheets = ['ORIGEN', 'SECTOR', 'SEXO', 'TIPO_PACIENTE', 'SI_NO', 'NACIONALIDAD', 'RESULTADO', 'de ENTIDADES', 'MUNICIPIOS']
catalogues = {}
for sheet in sheets:
  catalogue = pd.read_excel(nombre_cat, sheet_name='Catálogo ' + sheet)
  catalogue.columns = catalogue.columns.str.replace(' ', '_')
  catalogue.columns = catalogue.columns.str.replace('.', '')
  catalogue.columns = catalogue.columns.str.lower()
  new_catalogue = {sheet: catalogue}
  catalogues.update(new_catalogue)

In [37]:
catalogues

{'MUNICIPIOS':       clave_municipio        municipio  clave_entidad
 0                   1   AGUASCALIENTES              1
 1                   2         ASIENTOS              1
 2                   3         CALVILLO              1
 3                   4            COSÍO              1
 4                   5      JESÚS MARÍA              1
 ...               ...              ...            ...
 2496              999  NO ESPECIFICADO             32
 2497              999  NO ESPECIFICADO             36
 2498              997        NO APLICA             97
 2499              998        SE IGNORA             98
 2500              999  NO ESPECIFICADO             99
 
 [2501 rows x 3 columns], 'NACIONALIDAD':    clave      descripción
 0      1         MEXICANA
 1      2       EXTRANJERA
 2     99  NO ESPECIFICADO, 'ORIGEN':    clave      descripción
 0      1            USMER
 1      2   FUERA DE USMER
 2     99  NO ESPECIFICADO, 'RESULTADO':   unnamed:_0              unnamed:_1
 0    

In [38]:
catalogues['SEXO']

clave      descripción
0      1            MUJER
1      2           HOMBRE
2     99  NO ESPECIFICADO

##Perfilamiento

###Descripción General

In [39]:
# dimensión de los datos
print('Número de filas: {}'.format(data.shape[0]))
print('Número de columnas: {}'.format(data.shape[1]))

Número de filas: 1510311
Número de columnas: 35


In [40]:
# columnas (variables)
data.columns

Index(['FECHA_ACTUALIZACION', 'ID_REGISTRO', 'ORIGEN', 'SECTOR', 'ENTIDAD_UM',
       'SEXO', 'ENTIDAD_NAC', 'ENTIDAD_RES', 'MUNICIPIO_RES', 'TIPO_PACIENTE',
       'FECHA_INGRESO', 'FECHA_SINTOMAS', 'FECHA_DEF', 'INTUBADO', 'NEUMONIA',
       'EDAD', 'NACIONALIDAD', 'EMBARAZO', 'HABLA_LENGUA_INDIG', 'DIABETES',
       'EPOC', 'ASMA', 'INMUSUPR', 'HIPERTENSION', 'OTRA_COM',
       'CARDIOVASCULAR', 'OBESIDAD', 'RENAL_CRONICA', 'TABAQUISMO',
       'OTRO_CASO', 'RESULTADO', 'MIGRANTE', 'PAIS_NACIONALIDAD',
       'PAIS_ORIGEN', 'UCI'],
      dtype='object')

In [41]:
# limpieza del nombre de las columnas, remover espacios, carácteres especiales y pasar a minúsculas
data.columns = data.columns.str.replace(' ', '_')
data.columns = data.columns.str.replace('.', '')
data.columns = data.columns.str.lower()

In [42]:
# calcular número de valores nulos por columna
for col in data.columns:
  print('Valores nulos en "{}": {}'.format(col,data[col].isna().sum()))

Valores nulos en "fecha_actualizacion": 0
Valores nulos en "id_registro": 0
Valores nulos en "origen": 0
Valores nulos en "sector": 0
Valores nulos en "entidad_um": 0
Valores nulos en "sexo": 0
Valores nulos en "entidad_nac": 0
Valores nulos en "entidad_res": 0
Valores nulos en "municipio_res": 0
Valores nulos en "tipo_paciente": 0
Valores nulos en "fecha_ingreso": 0
Valores nulos en "fecha_sintomas": 0
Valores nulos en "fecha_def": 0
Valores nulos en "intubado": 0
Valores nulos en "neumonia": 0
Valores nulos en "edad": 0
Valores nulos en "nacionalidad": 0
Valores nulos en "embarazo": 0
Valores nulos en "habla_lengua_indig": 0
Valores nulos en "diabetes": 0
Valores nulos en "epoc": 0
Valores nulos en "asma": 0
Valores nulos en "inmusupr": 0
Valores nulos en "hipertension": 0
Valores nulos en "otra_com": 0
Valores nulos en "cardiovascular": 0
Valores nulos en "obesidad": 0
Valores nulos en "renal_cronica": 0
Valores nulos en "tabaquismo": 0
Valores nulos en "otro_caso": 0
Valores nulos 

In [43]:
# tipos de datos
data.dtypes

fecha_actualizacion    object
id_registro            object
origen                  int64
sector                  int64
entidad_um              int64
sexo                    int64
entidad_nac             int64
entidad_res             int64
municipio_res           int64
tipo_paciente           int64
fecha_ingreso          object
fecha_sintomas         object
fecha_def              object
intubado                int64
neumonia                int64
edad                    int64
nacionalidad            int64
embarazo                int64
habla_lengua_indig      int64
diabetes                int64
epoc                    int64
asma                    int64
inmusupr                int64
hipertension            int64
otra_com                int64
cardiovascular          int64
obesidad                int64
renal_cronica           int64
tabaquismo              int64
otro_caso               int64
resultado               int64
migrante                int64
pais_nacionalidad      object
pais_orige

###Limpieza

In [44]:
# convertimos a fecha algunas columnas
def cast_datetime_cols(df):
    timestamp_cols = [col for col in df.columns if "fecha" in col]
    df[timestamp_cols] = df[timestamp_cols].apply(lambda date_col: pd.to_datetime(date_col, errors="coerce"), axis=0)
    return df

data = cast_datetime_cols(data)

In [45]:
data.head()

fecha_actualizacion id_registro  origen  ...  pais_nacionalidad  pais_origen  uci
0          2020-09-12      114cd3       2  ...            México           99   97
1          2020-09-12      0bafd8       2  ...            México           99    2
2          2020-09-12      18747a       2  ...            México           99   97
3          2020-09-12      174174       2  ...            México           99   97
4          2020-09-12      0eae86       2  ...            México           99    1

[5 rows x 35 columns]

In [46]:
# unimos los datos con los catalogos para tener mejor entendimiento de ellos
data = (
    # unimos con el catálogo MUNICIPIOS y ENTIDADES
    data.merge(catalogues['de ENTIDADES'], how='left', left_on='entidad_res', right_on='clave_entidad').
    drop(columns=['entidad_res', 'abreviatura']).
    rename(columns={'entidad_federativa':'entidad_res'}).
    merge(catalogues['MUNICIPIOS'], how='left', left_on=['clave_entidad','municipio_res'], right_on=['clave_entidad','clave_municipio']).
    drop(columns=['municipio_res', 'clave_entidad', 'clave_municipio']).
    rename(columns={'municipio':'municipio_res'})
)

In [47]:
# unimos con el catálogo de ENTIDADES
entidades_cols = [col for col in data.columns if "entidad" in col and col != 'entidad_res']
for col in entidades_cols:
    data = (
        data.merge(catalogues['de ENTIDADES'], how='left', left_on=col, right_on='clave_entidad').
        drop(columns=[col, 'abreviatura', 'clave_entidad']).
        rename(columns={'entidad_federativa':col})
    )

In [48]:
# unimos con el catálogo SI_NO
sino_cols = ['intubado', 'neumonia', 'embarazo', 'habla_lengua_indig', 'diabetes', 'epoc', 'asma', 'inmusupr',
             'hipertension', 'otra_com', 'cardiovascular', 'obesidad', 'renal_cronica', 'tabaquismo', 
             'otro_caso', 'migrante', 'uci']
for col in sino_cols:
    data = (
        data.merge(catalogues['SI_NO'], how='left', left_on=col, right_on='clave').
        drop(columns=[col, 'clave']).
        rename(columns={'descripción':col})
    )

> **Nota:**
>
> Si ejecutaste una versión previa del cuaderno, notarás que en la siguiente sección aparecía un error. A continuación dejamos como un pequeño paréntesis un poco más de limpieza sobre el cuaderno.

1. Corregimos el nombre de las columnas:

In [49]:
catalogues['RESULTADO'].columns

Index(['unnamed:_0', 'unnamed:_1'], dtype='object')

In [50]:
catalogues['RESULTADO'].columns = ['clave', 'descripción']
catalogues['RESULTADO'].columns

Index(['clave', 'descripción'], dtype='object')

2. Removemos la (primera) fila que no nos es útil.

In [51]:
catalogues['RESULTADO'] = catalogues['RESULTADO'].drop(index=0)

3. Convertimos el tipo de dato.

In [52]:
catalogues['RESULTADO']['clave'] = catalogues['RESULTADO']['clave'].astype(int)

> Fin del paréntesis. Hasta esta sección, esta parte de limpieza permite continuar con el proceso de manera normal.

In [53]:
# unimos otras columnnas que comparten lógica
other_cols = ['origen', 'sector', 'sexo', 'tipo_paciente', 'nacionalidad', 'resultado']
for col, cat in zip(other_cols, [x.upper() for x in other_cols]):
    data = (
        data.merge(catalogues[cat], how='left', left_on=col, right_on='clave').
        drop(columns=[col, 'clave']).
        rename(columns={'descripción':col})
    )

In [54]:
# creamos algunas variables que podrían ser útiles después
data['defuncion'] = ~data.fecha_def.isna()
data['mes_def'] = data.fecha_def.dt.month
data['año_def'] = data.fecha_def.dt.year
data['mes_ingreso'] = data.fecha_ingreso.dt.month
data['año_ingreso'] = data.fecha_ingreso.dt.year

In [55]:
data.head()

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-12      114cd3  ...           7        2020
1          2020-09-12      0bafd8  ...           4        2020
2          2020-09-12      18747a  ...           4        2020
3          2020-09-12      174174  ...           4        2020
4          2020-09-12      0eae86  ...           4        2020

[5 rows x 40 columns]

###Filtros

**Distintas formas de filtrar lo mismo**

In [56]:
# usando query
data.query('nacionalidad == "MEXICANA"').head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-12      114cd3  ...           7        2020
1          2020-09-12      0bafd8  ...           4        2020

[2 rows x 40 columns]

In [57]:
# haciendo referencia a la columna como índice
data[data['nacionalidad'] == 'MEXICANA'].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-12      114cd3  ...           7        2020
1          2020-09-12      0bafd8  ...           4        2020

[2 rows x 40 columns]

In [58]:
# seleccionando directamente la columna que queremos filtrar
data[data.nacionalidad == 'MEXICANA'].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-12      114cd3  ...           7        2020
1          2020-09-12      0bafd8  ...           4        2020

[2 rows x 40 columns]

**Filtrado múltiple**

In [59]:
# usando query podemos poner en el string del query todas las condiciones que queramos
data.query('nacionalidad == "MEXICANA" & entidad_res == "CIUDAD DE MÉXICO"').head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
3          2020-09-12      174174  ...           4        2020
6          2020-09-12      1a73c1  ...           5        2020

[2 rows x 40 columns]

In [60]:
# usando el filtrado tradicional debemos encerrar en paréntesis cada condición que queremos que cumpla nuestro filtro
data[(data['nacionalidad'] == "MEXICANA") & (data.entidad_res == "CIUDAD DE MÉXICO")].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
3          2020-09-12      174174  ...           4        2020
6          2020-09-12      1a73c1  ...           5        2020

[2 rows x 40 columns]

###Resúmenes

In [61]:
# una columna, una métrica
data.groupby('sexo').edad.mean()

sexo
HOMBRE    42.667259
MUJER     41.639353
Name: edad, dtype: float64

In [62]:
# una columna, varias métricas
data.groupby('sexo').agg({'edad': ['min','mean','max']})

edad                
        min       mean  max
sexo                       
HOMBRE    0  42.667259  120
MUJER     0  41.639353  120

In [63]:
# varias columnas, una métrica
data.groupby('sexo').agg({'edad': 'mean', 'mes_ingreso': 'mean', 'mes_def': 'mean'})

edad  mes_ingreso   mes_def
sexo                                    
HOMBRE  42.667259     6.730779  6.497591
MUJER   41.639353     6.807770  6.572409

In [64]:
# varias columnas, varias métricas
data.groupby('sexo').agg({'edad': ['min','mean','max'], 'mes_ingreso': ['min','mean','max'], 'mes_def': ['min','mean','max']})

edad                 mes_ingreso               mes_def               
        min       mean  max         min      mean max     min      mean  max
sexo                                                                        
HOMBRE    0  42.667259  120           1  6.730779   9     1.0  6.497591  9.0
MUJER     0  41.639353  120           1  6.807770   9     1.0  6.572409  9.0

###Descripción por Tipo

In [65]:
# Filtrar por tipos generales (categóricas y numéricas)
numerical = data.select_dtypes(include='number')
categorical = data.select_dtypes(include=['object', 'bool', 'category', 'datetime64[ns]'])

In [66]:
numerical.head()

edad  mes_def  año_def  mes_ingreso  año_ingreso
0    52      NaN      NaN            7         2020
1    57      5.0   2020.0            4         2020
2    44      NaN      NaN            4         2020
3    28      NaN      NaN            4         2020
4    53      NaN      NaN            4         2020

In [67]:
categorical.head()

fecha_actualizacion id_registro  ...            resultado defuncion
0          2020-09-12      114cd3  ...  Positivo SARS-CoV-2     False
1          2020-09-12      0bafd8  ...  Positivo SARS-CoV-2      True
2          2020-09-12      18747a  ...  Positivo SARS-CoV-2     False
3          2020-09-12      174174  ...  Positivo SARS-CoV-2     False
4          2020-09-12      0eae86  ...  Positivo SARS-CoV-2     False

[5 rows x 35 columns]

In [68]:
# para conocer todos los tipo de datos
?data.select_dtypes

####Categóricas

In [69]:
# Cardinalidad - Valores únicos por columna
for col in categorical.columns:
  print('Valores únicos en "{}": {}'.format(col,categorical[col].nunique()))

Valores únicos en "fecha_actualizacion": 1
Valores únicos en "id_registro": 1510311
Valores únicos en "fecha_ingreso": 256
Valores únicos en "fecha_sintomas": 256
Valores únicos en "fecha_def": 201
Valores únicos en "pais_nacionalidad": 119
Valores únicos en "pais_origen": 77
Valores únicos en "entidad_res": 32
Valores únicos en "municipio_res": 2164
Valores únicos en "entidad_um": 32
Valores únicos en "entidad_nac": 33
Valores únicos en "intubado": 4
Valores únicos en "neumonia": 3
Valores únicos en "embarazo": 4
Valores únicos en "habla_lengua_indig": 3
Valores únicos en "diabetes": 3
Valores únicos en "epoc": 3
Valores únicos en "asma": 3
Valores únicos en "inmusupr": 3
Valores únicos en "hipertension": 3
Valores únicos en "otra_com": 3
Valores únicos en "cardiovascular": 3
Valores únicos en "obesidad": 3
Valores únicos en "renal_cronica": 3
Valores únicos en "tabaquismo": 3
Valores únicos en "otro_caso": 3
Valores únicos en "migrante": 3
Valores únicos en "uci": 4
Valores únicos en

In [70]:
# Unicidad - proporción de valores únicos por columna
total = data.shape[0]
for col in categorical.columns:
  print('Porcentaje de valores únicos en "{}": {}%'.format(col,round(categorical[col].nunique()/total*100,2)))

Porcentaje de valores únicos en "fecha_actualizacion": 0.0%
Porcentaje de valores únicos en "id_registro": 100.0%
Porcentaje de valores únicos en "fecha_ingreso": 0.02%
Porcentaje de valores únicos en "fecha_sintomas": 0.02%
Porcentaje de valores únicos en "fecha_def": 0.01%
Porcentaje de valores únicos en "pais_nacionalidad": 0.01%
Porcentaje de valores únicos en "pais_origen": 0.01%
Porcentaje de valores únicos en "entidad_res": 0.0%
Porcentaje de valores únicos en "municipio_res": 0.14%
Porcentaje de valores únicos en "entidad_um": 0.0%
Porcentaje de valores únicos en "entidad_nac": 0.0%
Porcentaje de valores únicos en "intubado": 0.0%
Porcentaje de valores únicos en "neumonia": 0.0%
Porcentaje de valores únicos en "embarazo": 0.0%
Porcentaje de valores únicos en "habla_lengua_indig": 0.0%
Porcentaje de valores únicos en "diabetes": 0.0%
Porcentaje de valores únicos en "epoc": 0.0%
Porcentaje de valores únicos en "asma": 0.0%
Porcentaje de valores únicos en "inmusupr": 0.0%
Porcenta

In [71]:
# Moda - valor que más se repite en cada columna excepto los indicadores únicos
for col in categorical.drop(columns='id_registro').columns:
  print('Valor más popular en "{}": {}'.format(col, statistics.mode(categorical[col])))

Valor más popular en "fecha_actualizacion": 2020-09-12 00:00:00
Valor más popular en "fecha_ingreso": 2020-07-20 00:00:00
Valor más popular en "fecha_sintomas": 2020-07-20 00:00:00
Valor más popular en "fecha_def": NaT
Valor más popular en "pais_nacionalidad": México
Valor más popular en "pais_origen": 99
Valor más popular en "entidad_res": CIUDAD DE MÉXICO
Valor más popular en "municipio_res": IZTAPALAPA
Valor más popular en "entidad_um": CIUDAD DE MÉXICO
Valor más popular en "entidad_nac": CIUDAD DE MÉXICO
Valor más popular en "intubado": NO APLICA
Valor más popular en "neumonia": NO 
Valor más popular en "embarazo": NO 
Valor más popular en "habla_lengua_indig": NO 
Valor más popular en "diabetes": NO 
Valor más popular en "epoc": NO 
Valor más popular en "asma": NO 
Valor más popular en "inmusupr": NO 
Valor más popular en "hipertension": NO 
Valor más popular en "otra_com": NO 
Valor más popular en "cardiovascular": NO 
Valor más popular en "obesidad": NO 
Valor más popular en "r

####Numéricas

In [72]:
numerical.describe()

edad       mes_def  año_def   mes_ingreso  año_ingreso
count  1.510311e+06  90195.000000  90195.0  1.510311e+06    1510311.0
mean   4.214413e+01      6.524896   2020.0  6.769961e+00       2020.0
std    1.680812e+01      1.306444      0.0  1.395983e+00          0.0
min    0.000000e+00      1.000000   2020.0  1.000000e+00       2020.0
25%    3.000000e+01      6.000000   2020.0  6.000000e+00       2020.0
50%    4.100000e+01      7.000000   2020.0  7.000000e+00       2020.0
75%    5.300000e+01      8.000000   2020.0  8.000000e+00       2020.0
max    1.200000e+02      9.000000   2020.0  9.000000e+00       2020.0

##Análisis Exploratorio

En esta sección responderemos algunas preguntas sobre el set de datos usando las herramientas que hemos visto hasta ahora

In [73]:
# ¿Cuántas pruebas por origen se realizan?
data.groupby('origen').id_registro.count()

origen
FUERA DE USMER    1010999
USMER              499312
Name: id_registro, dtype: int64

In [74]:
data.origen.value_counts()

FUERA DE USMER    1010999
USMER              499312
Name: origen, dtype: int64

In [75]:
# ¿Cuántas pruebas por sector se realizan?
data.sector.value_counts()

SSA                942672
IMSS               386225
ISSSTE              57072
PRIVADA             55614
ESTATAL             27163
PEMEX               12919
IMSS-BIENESTAR      10158
SEMAR                7796
SEDENA               7371
MUNICIPAL            1201
UNIVERSITARIO        1046
DIF                   843
CRUZ ROJA             223
NO ESPECIFICADO         8
Name: sector, dtype: int64

In [76]:
# ¿Cuántos pacientes por sexo tenemos?
data.sexo.value_counts()

MUJER     768635
HOMBRE    741676
Name: sexo, dtype: int64

In [77]:
# ¿Cuál es la proporción de los pacientes por sexo?
data.sexo.value_counts(normalize=True)

MUJER     0.508925
HOMBRE    0.491075
Name: sexo, dtype: float64

In [78]:
# ¿Cuál es la proporción de resultados de las pruebas?
data.resultado.value_counts(normalize=True)

No positivo SARS-CoV-2    0.502670
Positivo SARS-CoV-2       0.439627
Resultado pendiente       0.057703
Name: resultado, dtype: float64

In [79]:
# ¿Cuántas pruebas positivas tenemos en el set?
data[data.resultado == 'Positivo SARS-CoV-2'].id_registro.count()

663973

In [80]:
# ¿Cuántos casos positivos tenemos por mes?
data[data.resultado == 'Positivo SARS-CoV-2'].mes_ingreso.value_counts().sort_index()

1         2
2         8
3      2530
4     26567
5     86557
6    152980
7    199471
8    155971
9     39887
Name: mes_ingreso, dtype: int64

In [81]:
# ¿Cuál es el resultado de las pruebas por sexo?
data.groupby(['sexo','resultado'])['id_registro'].count()

sexo    resultado             
HOMBRE  No positivo SARS-CoV-2    352619
        Positivo SARS-CoV-2       346464
        Resultado pendiente        42593
MUJER   No positivo SARS-CoV-2    406569
        Positivo SARS-CoV-2       317509
        Resultado pendiente        44557
Name: id_registro, dtype: int64

In [82]:
data.groupby('sexo').resultado.value_counts()

sexo    resultado             
HOMBRE  No positivo SARS-CoV-2    352619
        Positivo SARS-CoV-2       346464
        Resultado pendiente        42593
MUJER   No positivo SARS-CoV-2    406569
        Positivo SARS-CoV-2       317509
        Resultado pendiente        44557
Name: resultado, dtype: int64

In [83]:
# ¿Qué proporción de resultados de pruebas tenemos por sexo?
data.groupby('sexo').resultado.value_counts(normalize=True)

sexo    resultado             
HOMBRE  No positivo SARS-CoV-2    0.475435
        Positivo SARS-CoV-2       0.467137
        Resultado pendiente       0.057428
MUJER   No positivo SARS-CoV-2    0.528949
        Positivo SARS-CoV-2       0.413082
        Resultado pendiente       0.057969
Name: resultado, dtype: float64

In [84]:
# ¿Cuál es la edad de los pacientes por sexo?
data.groupby('sexo').edad.describe()

count       mean        std  min   25%   50%   75%    max
sexo                                                                
HOMBRE  741676.0  42.667259  17.126286  0.0  30.0  41.0  54.0  120.0
MUJER   768635.0  41.639353  16.479581  0.0  30.0  40.0  52.0  120.0

In [85]:
def q25(x):
  return x.quantile(0.25)

def q75(x):
  return x.quantile(0.75)

data.groupby('sexo').agg({'edad': ['min', q25, 'median', q75, 'max']})

edad                    
        min q25 median q75  max
sexo                           
HOMBRE    0  30     41  54  120
MUJER     0  30     40  52  120

In [86]:
# ¿Cuál es la edad de los pacientes dependiendo de si son intubados o no?
data.groupby('intubado').edad.describe()

count       mean        std  min   25%   50%   75%    max
intubado                                                                      
NO                221572.0  53.304786  19.684371  0.0  42.0  55.0  67.0  118.0
NO APLICA        1248894.0  39.686951  15.079782  0.0  29.0  38.0  49.0  120.0
NO ESPECIFICADO      194.0  54.443299  16.817502  0.0  45.0  57.0  65.0   92.0
SI                 39651.0  57.111826  17.914277  0.0  48.0  59.0  69.0  108.0

In [87]:
data.groupby('intubado').agg({'edad': ['min', q25, 'median', q75, 'max']})

edad                    
                 min q25 median q75  max
intubado                                
NO                 0  42     55  67  118
NO APLICA          0  29     38  49  120
NO ESPECIFICADO    0  45     57  65   92
SI                 0  48     59  69  108

In [88]:
# ¿Cuál es la edad máxima y cuál es la mínima de los pacientes fallecidos?
edad_def_max = data[data.defuncion].edad.max()
data[data.defuncion & (data.edad == edad_def_max)]

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
581442          2020-09-12      1b7b20  ...           5        2020

[1 rows x 40 columns]

In [89]:
edad_def_min = data[data.defuncion].edad.min()
data[data.defuncion & (data.edad == edad_def_min)]

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
11277            2020-09-12      18feee  ...           6        2020
11565            2020-09-12      083c34  ...           5        2020
26116            2020-09-12      02f8e2  ...           5        2020
50494            2020-09-12      014b5d  ...           5        2020
60925            2020-09-12      053398  ...           6        2020
...                     ...         ...  ...         ...         ...
1439056          2020-09-12      0860f4  ...           6        2020
1462455          2020-09-12      1e94e4  ...           8        2020
1473848          2020-09-12      00f39c  ...           7        2020
1487374          2020-09-12      07cf0b  ...           8        2020
1502388          2020-09-12      3b8095  ...           9        2020

[252 rows x 40 columns]

In [90]:
# ¿Cuál es la edad de los pacientes que fallecen?
data.groupby('defuncion').edad.describe()

count       mean        std  min   25%   50%   75%    max
defuncion                                                               
False      1420116.0  40.888838  16.104672  0.0  29.0  39.0  51.0  120.0
True         90195.0  61.908676  15.218544  0.0  53.0  63.0  73.0  109.0

In [91]:
# ¿Cuántas defunciones tenemos por mes?
data[data.defuncion].mes_def.value_counts().sort_index()

1.0        6
2.0       12
3.0      232
4.0     5294
5.0    16091
6.0    21461
7.0    23855
8.0    18791
9.0     4453
Name: mes_def, dtype: int64

In [92]:
# ¿Cuál es la proporción de diábeticos para las defunciones?
data.groupby('diabetes').defuncion.value_counts(normalize=True)

diabetes   defuncion
NO         False        0.958145
           True         0.041855
SE IGNORA  False        0.873868
           True         0.126132
SI         False        0.813339
           True         0.186661
Name: defuncion, dtype: float64

In [93]:
# ¿Cuál es la proporción de asmáticos para las defunciones?
data.groupby('asma').defuncion.value_counts(normalize=True)

asma       defuncion
NO         False        0.939952
           True         0.060048
SE IGNORA  False        0.867149
           True         0.132851
SI         False        0.958165
           True         0.041835
Name: defuncion, dtype: float64

TAREA 3

1. ¿Tienen los pacientes con hipertensión un riesgo más alto de defunción?

In [94]:
data.groupby('hipertension').defuncion.value_counts(normalize=True)

hipertension  defuncion
NO            False        0.960768
              True         0.039232
SE IGNORA     False        0.873276
              True         0.126724
SI            False        0.834452
              True         0.165548
Name: defuncion, dtype: float64

Solución -> Sí, ya que el 16.55% con tal padecimiento resultaron enfermos (mientras que sólo el 3.92% de los que no tenian la enfermedad se vieron contagiados)

2. ¿Cuántos casos confirmados se tienen por Estado?

In [95]:
data[data.resultado == 'Positivo SARS-CoV-2'].entidad_res.value_counts()

CIUDAD DE MÉXICO                   110809
MÉXICO                              73353
GUANAJUATO                          36391
NUEVO LEÓN                          33510
VERACRUZ DE IGNACIO DE LA LLAVE     30788
TABASCO                             29911
PUEBLA                              29046
TAMAULIPAS                          26778
COAHUILA DE ZARAGOZA                23906
SONORA                              23057
JALISCO                             22997
SAN LUIS POTOSÍ                     20585
BAJA CALIFORNIA                     18299
MICHOACÁN DE OCAMPO                 17474
SINALOA                             17289
YUCATÁN                             16453
GUERRERO                            16354
OAXACA                              14726
HIDALGO                             11364
QUINTANA ROO                        11055
CHIHUAHUA                            8945
BAJA CALIFORNIA SUR                  8871
DURANGO                              7628
QUERÉTARO                         

3. ¿Cuántas defunciones se tienen por Estado?

In [96]:
data[data.defuncion == True]. entidad_res.value_counts()

MÉXICO                             13635
CIUDAD DE MÉXICO                   12012
VERACRUZ DE IGNACIO DE LA LLAVE     4853
PUEBLA                              4635
BAJA CALIFORNIA                     4081
JALISCO                             3899
SINALOA                             3751
NUEVO LEÓN                          3554
SONORA                              3282
GUANAJUATO                          3185
TABASCO                             2929
TAMAULIPAS                          2852
GUERRERO                            2303
COAHUILA DE ZARAGOZA                2196
HIDALGO                             2078
CHIHUAHUA                           2043
MICHOACÁN DE OCAMPO                 1929
QUINTANA ROO                        1915
YUCATÁN                             1821
SAN LUIS POTOSÍ                     1785
OAXACA                              1622
TLAXCALA                            1331
MORELOS                             1255
CHIAPAS                             1221
QUERÉTARO       

4. ¿Cuántos fallecimientos han ocurrido en el Estado con mayor número de casos confirmados?

In [97]:
data[data.defuncion & (data.entidad_res == "CIUDAD DE MÉXICO")].defuncion.value_counts()

True    12012
Name: defuncion, dtype: int64

5. ¿Cuántos fallecimientos han ocurrido en los pacientes Ambulatorios?

In [98]:
data[data.defuncion & (data.tipo_paciente == "AMBULATORIO")].defuncion.value_counts()

True    9609
Name: defuncion, dtype: int64